File: chomp.c
#include <stdio.h>

int main(){

	int testCases, caseOn;
	
	//Read in the number of cases
	scanf("%d", &testCases);
	
	for(caseOn = 1;caseOn<=testCases;caseOn++){
		
		int n, k, i, j;
		
		
		scanf("%d", &n);
		scanf("%d", &k);
		
		//Read in the food values
		int food[n];
		for(i = 0;i<n;i++){
			scanf("%d", &food[i]);
		}
		
		
		//This array stores the sum of the food values from 0 to i in position i
		//This allows us to quickly get the sum of any range of the food array
		//The sum from position i to position j in the food array is equal to
		//prefixSums[j]-prefixSums[i-1] which is the sum from 0 to j minus the 
		//sum from 0 to i-1
		long long prefixSums[n];
		for(i = 0;i<n;i++){
			prefixSums[i] = food[i];
			if(i > 0){
				prefixSums[i]+=prefixSums[i-1];
			}
		}
		
		long long answer = 0;
		
		//This array keeps track of the frequency of the sums of all possible
		//first chomps that end at least two positions before th
		long long counts[k];
		
		for(i = 0;i<k;i++){
			counts[i] = 0;
		}
		
		
		//We are looping over the start position of the second chomp and using
		//counts array to keep track of possibilites for the first chomp
		for(i = 2; i<n;i++){
		
			//We need to add in all the first chomps that end at position i-2
			//as this is now a valid position for them to end. To do this we loop
			//over all possible start positions of the first chomp and add the
			//sum from j to i-2 mod k to the counts array
			for(j = 0;j<=i-2;j++){
				long long sum = prefixSums[i-2]-(j == 0 ? 0 : prefixSums[j-1]);
				sum%=k;
				counts[sum]++;
			}
			
			//This loops over all possible end positions of the second chomp
			for(j = i;j<n;j++){
			
				long long sum = prefixSums[j]-(i == 0 ? 0 : prefixSums[i-1]);
				sum%=k;
				
				//We need the sum of the two chomps to be equal to 0 when moded
				//by k in order for the sum of the remaining food to be the same
				//mod k as the original plate
				int need = (k-sum)%k;
				
				//Add the number of first chomps that have the right sum to our
				//answer
				answer+=counts[need];
			}
			
		}
		
		
		
		printf("Plate #%d: %lld\n", caseOn, answer);
	}

	return 0;
}

File: chomp.java
import java.util.*;
import java.io.*;

public class chomp {
	static int[] prefixSums;
	static int k;
	public static void main(String[] args) throws IOException {
		
		Scanner in = new Scanner(System.in);
		PrintWriter out = new PrintWriter(System.out);
		int p = in.nextInt();
		
		for(int plate = 1; plate <= p; plate++){
			
			//variables from the input
			int n = in.nextInt();
			k = in.nextInt();
			
			//this are all the food items
			int[] a = new int[n];
			for (int i = 0; i < a.length; i++) {
				a[i] = in.nextInt();
			}
			
			
			//Since after the chomps the sum of all the items needs to the same (mod k),
			//the sum of the items with the sums together need to 0 (mod k)
			//This approaches finds all such pairs of intervals in O(n^2) time complexity 
			//The chomps of the problem are referred to as intervals
			
			//prefixSums[i] holds the sum of all a[j] s.t. 0 <= j <= i
			//It is used for the sumOverRange function
			prefixSums = new int[n];
			prefixSums[0] = a[0];
			for (int i = 1; i < a.length; i++) {
				prefixSums[i] = (prefixSums[i-1] + a[i])%k;
			}
			long answer = 0;
			
			//canMakeWithLeft[i] is The number of ways to make a interval which 
			//sums to i (mod k) such that the end point of the interval is 2 or more less than startOfRight
			int[] canMakeWithLeft = new int[k];
			
			//The start of the right interval can not be less than 2, as there needs to be at least one index for the
			//left interval and one index to separate them
			for(int startOfRight = 2; startOfRight < n; startOfRight++){
				int endOfLeft = startOfRight-2;
				//update canMakeWithLeft with the new end point that can used for the left intervals
				for(int startOfLeft = 0; startOfLeft <= endOfLeft; startOfLeft++){
					int sumLeft = sumOverRange(startOfLeft,endOfLeft,a);
					canMakeWithLeft[sumLeft]++;
				}
				
				for(int endOfRight = startOfRight; endOfRight < n; endOfRight++){
					int sumRight = sumOverRange(startOfRight,endOfRight,a);
					//I'm look for the sumLeft such that sumRight+sumLeft is 0 mod k
					//Therefore the sumLeft is (k - sumRight) mod k
					answer += canMakeWithLeft[(k-sumRight)%k];
				}
				
				
			}
			
			out.println("Plate #"+plate+": " + answer);
			
		}
		in.close();
		out.close();
	}
	
	//returns the sum (mod k) of all a[i] s.t. start <= i <= end.
	static int sumOverRange(int start, int end, int[] a){
		
		int sum =  prefixSums[end] - (prefixSums[start]-a[start]);
		sum %= k;
		if(sum < 0) sum += k;
		return sum;
	}

}
